File: packages/contracts/src/messaging/forecast.ts
import { z } from 'zod';
import { ObjectUri } from './envelope.js';
import { AccuracyMetric, ModelPackage } from '../http/scenarios.js';

/**
 * Schemas de payload de `forecast.request` e `forecast.result`.
 *
 * O motor recebe rótulos, não ids do banco — ele não conhece o schema Prisma
 * (Princípio III). Campos numéricos aqui são cardinalidades (meses, contagens),
 * nunca grandezas monetárias, portanto z.int() é correto (Princípio V).
 *
 * Ver: specs/001-sop-cycle-forecasting/contracts/messaging.md
 */

const uuid = z.string().uuid();

export const ForecastRequestParams = z.object({
  groupingLabels: z.array(z.string()).min(1),
  granularLabels: z.array(z.string()).min(1),
  prorationMonths: z.int().min(1),
  horizonMonths: z.int().min(1),
  accuracyMetric: AccuracyMetric,
  modelPackage: ModelPackage,
  decimalScale: z.int().min(1).max(10),
  nonNegativeForecast: z.boolean(),
});
export type ForecastRequestParams = z.infer<typeof ForecastRequestParams>;

export const ForecastRequestPayload = z.object({
  jobId: uuid,
  scenarioId: uuid,
  inputUri: ObjectUri,
  outputPrefix: ObjectUri,
  modelCatalogVersion: z.string().min(1),
  params: ForecastRequestParams,
});
export type ForecastRequestPayload = z.infer<typeof ForecastRequestPayload>;

const ForecastStats = z.object({
  seriesCount: z.int().min(0),
  itemCount: z.int().min(0),
  durationMs: z.int().min(0),
});

const ForecastFailure = z.object({
  code: z.string().min(1),
  message: z.string(),
});

const ForecastResultCompleted = z.object({
  jobId: uuid,
  status: z.literal('completed'),
  outputUri: ObjectUri,
  seriesUri: ObjectUri,
  modelCatalogVersion: z.string().min(1),
  stats: ForecastStats,
  failure: z.null(),
});

const ForecastResultFailed = z.object({
  jobId: uuid,
  status: z.literal('failed'),
  outputUri: z.null(),
  seriesUri: z.null(),
  modelCatalogVersion: z.string().min(1),
  stats: z.null(),
  failure: ForecastFailure,
});

export const ForecastResultPayload = z.discriminatedUnion('status', [
  ForecastResultCompleted,
  ForecastResultFailed,
]);
export type ForecastResultPayload = z.infer<typeof ForecastResultPayload>;

File: services/forecast-engine/src/forecast_engine/messaging/forecast.py
"""Schemas de payload de forecast.request e forecast.result — espelha os schemas Zod.

Verificado contra os MESMOS vetores dourados que o lado TypeScript.

Ver: specs/001-sop-cycle-forecasting/contracts/messaging.md
"""

from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, field_validator


class ForecastRequestParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    groupingLabels: list[str]  # noqa: N815
    granularLabels: list[str]  # noqa: N815
    prorationMonths: int  # noqa: N815
    horizonMonths: int  # noqa: N815
    accuracyMetric: Literal["WMAPE", "MAPE", "BIAS"]  # noqa: N815
    modelPackage: Literal["FAST", "STANDARD", "COMPLETE"]  # noqa: N815
    decimalScale: int  # noqa: N815
    nonNegativeForecast: bool  # noqa: N815

    @field_validator("groupingLabels", "granularLabels")
    @classmethod
    def _nao_vazio(cls, v: list[str]) -> list[str]:
        if not v:
            raise ValueError("lista de rótulos não pode ser vazia")
        return v

    @field_validator("prorationMonths", "horizonMonths")
    @classmethod
    def _positivo(cls, v: int) -> int:
        if v < 1:
            raise ValueError("deve ser pelo menos 1")
        return v

    @field_validator("decimalScale")
    @classmethod
    def _escala_valida(cls, v: int) -> int:
        if not 1 <= v <= 10:
            raise ValueError("decimalScale deve estar entre 1 e 10")
        return v


class ForecastRequestPayload(BaseModel):
    """Payload de `forecast.request` — o motor recebe rótulos, não ids do banco (Princípio III)."""

    model_config = ConfigDict(extra="forbid")

    jobId: UUID  # noqa: N815
    scenarioId: UUID  # noqa: N815
    inputUri: str  # noqa: N815
    outputPrefix: str  # noqa: N815
    modelCatalogVersion: str  # noqa: N815
    params: ForecastRequestParams

    @field_validator("inputUri", "outputPrefix")
    @classmethod
    def _s3_uri(cls, v: str) -> str:
        import re
        if not re.match(r"^s3://[a-z0-9.\-]+/.+$", v):
            raise ValueError(f"esperado um URI s3://bucket/caminho, recebido {v!r}")
        return v


class ForecastStats(BaseModel):
    model_config = ConfigDict(extra="forbid")

    seriesCount: int  # noqa: N815
    itemCount: int  # noqa: N815
    durationMs: int  # noqa: N815


class ForecastFailure(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    message: str


class ForecastResultCompleted(BaseModel):
    model_config = ConfigDict(extra="forbid")

    jobId: UUID  # noqa: N815
    status: Literal["completed"]
    outputUri: str  # noqa: N815
    seriesUri: str  # noqa: N815
    modelCatalogVersion: str  # noqa: N815
    stats: ForecastStats
    failure: None

    @field_validator("outputUri", "seriesUri")
    @classmethod
    def _s3_uri(cls, v: str) -> str:
        import re
        if not re.match(r"^s3://[a-z0-9.\-]+/.+$", v):
            raise ValueError(f"esperado um URI s3://bucket/caminho, recebido {v!r}")
        return v


class ForecastResultFailed(BaseModel):
    model_config = ConfigDict(extra="forbid")

    jobId: UUID  # noqa: N815
    status: Literal["failed"]
    outputUri: None  # noqa: N815
    seriesUri: None  # noqa: N815
    modelCatalogVersion: str  # noqa: N815
    stats: None
    failure: ForecastFailure


def parse_forecast_result(data: dict) -> ForecastResultCompleted | ForecastResultFailed:
    """Analisa o payload de resultado discriminando pelo campo `status`."""
    status = data.get("status")
    if status == "completed":
        return ForecastResultCompleted.model_validate(data)
    if status == "failed":
        return ForecastResultFailed.model_validate(data)
    raise ValueError(f"status inválido: {status!r}; esperado 'completed' ou 'failed'")
